commit 794273c2729f4b79ddd321d3fa90f9b300e86800
Author: Vincent Bernat <bernat@luffy.cx>
Date:   Mon Dec 5 14:47:14 2011 +0100

    Don't use bind() with AF_UNSPEC
    
    tcp_bind_connect() may be called with a non NULL bind_addr variable
    but filled with 0. This cause bind() to be called with AF_UNSPEC and
    make it fail with recent kernels.

diff --git a/keepalived/core/layer4.c b/keepalived/core/layer4.c
index 5725500..5b33417 100644
--- a/keepalived/core/layer4.c
+++ b/keepalived/core/layer4.c
@@ -43,7 +43,7 @@ tcp_bind_connect(int fd, struct sockaddr_storage *addr, struct sockaddr_storage
 	fcntl(fd, F_SETFL, val | O_NONBLOCK);
 
 	/* Bind socket */
-	if (bind_addr) {
+	if (bind_addr && ((struct sockaddr *) bind_addr)->sa_family != AF_UNSPEC) {
 		addrlen = sizeof(*bind_addr);
 		if (bind(fd, (struct sockaddr *) bind_addr, addrlen) != 0)
 			return connect_error;
